drivers/serial/uart_nrfx_uarte.c

```diff
@@ -915,33 +931,31 @@ static void uarte_nrfx_poll_out(struct device *dev, unsigned char c)
 				uarte_nrfx_isr_async(dev);
 			}
 		}
-		/* Set tx_size but don't set tx_buf to differentiate this
-		 * transmission from the one started with uarte_nrfx_tx,
-		 * this way uarte_nrfx_tx will return -EBUSY and poll out will
-		 * work when interrupted.
+		/* Use tx_size as lock, this way uarte_nrfx_tx will
+		 * return -EBUSY during poll_out.
 		 */
-		data->async->tx_size = 1;
-		nrf_uarte_int_disable(uarte,
-				      NRF_UARTE_INT_ENDTX_MASK |
-				      NRF_UARTE_INT_TXSTOPPED_MASK);
-	}
+		lock = &data->async->tx_size;
```

Comment:
There is a chance that once transfer will be completed a new one will come. In that case, if in interrupt we overwrite length (*lock = 1).  Maybe we should lock immediately in while loop which is doing the check? Something like:
```
do {
  while(txbuf) {  uarte_nrfx_isr_async(dev); }
} while (atomic_cas(&lock, 0, 1) == false);
```

Reply:
If a new one comes it will be handled normally, in worst case scenario we will lose first byte send by poll_out